crates/mdb/src/lib.rs

```diff
@@ -196,7 +196,7 @@ impl Deferred {
 
 pub struct SubscriberConfig {
     ptr: *mut mdb_sys::mdb_subscriber_config_t,
-    _on_message: Deferred,
+    on_message: Option<Deferred>,
```

Comment:
`Option<Deferred>` is kind of like a Unique pointer now, it's an interesting way to move something out of a struct. Does it make sense to implement `take` on Deferred instead to "transfer ownership" of the pointer?

Reply:
Overall, I think this pattern is a little bit awkward, but it's the best one I have found. Another way to express what I think we want to do is the below. Kind of. The added generic argument is annoying and makes it impossible to reuse the pattern in axevent where multiple callbacks are stored on the same struct, and it has UB.

```rust
pub struct SubscriberConfig<F> {
    ptr: *mut mdb_sys::mdb_subscriber_config_t,
    on_message: Option<Box<F>>,
}

impl<F> SubscriberConfig<F>
where
    F: for<'a> FnMut(Message<'a>) + Send + 'static,
{
    pub fn try_new(topic: &CStr, source: &CStr, on_message: F) -> Result<Self, Error> {
        debug!("Creating {}...", any::type_name::<Self>());
        unsafe {
            let mut on_message = Box::new(on_message);
            let mut error: *mut mdb_sys::mdb_error_t = std::ptr::null_mut();
            let ptr = mdb_sys::mdb_subscriber_config_create(
                topic.as_ptr(),
                source.as_ptr(),
                Some(Self::on_message::<F>),
                &mut *on_message as *mut F as *mut c_void,
                &mut error,
            );
            match (ptr.is_null(), error.is_null()) {
                (false, false) => {
                    panic!("mdb_subscriber_config_create returned both a connection and an error")
                }
                (false, true) => Ok(Self { ptr, on_message: Some(on_message) }),
                (true, false) => Err(Error::new_owned(error)),
                (true, true) => panic!(
                    "mdb_subscriber_config_create returned neither a connection nor an error"
                ),
            }
        }
    }
```

Reply:
I agree it's kind of awkward, what I meant was:

```
impl Drop for Deferred {
    fn drop(&mut self) {
        if let Some(inner) = self.0.take() {
            inner();
        }
    }
}
impl Deferred {
    ...
    fn take(&mut self) -> Self {
        Self (self.0.take())
    }
}
```

Then there would be no need to wrap the `Deferred` in an `Option`. It could still be moved, only this time it's moved out of the Deferred, not the struct that owns it.

Reply:
I see what you mean. It's neat but I'm concerned it will make it more difficult for future maintainers to follow the callback through the code; with the `unwrap` it is clear that the author (us) intended for the closure to still be present when we `take` it from the config. But I could go either way and I have implemented your suggestion with a slight modification (using `mem::take` instead of implementing `fn take(&mut self)`.

Reply:
> I'm concerned it will make it more difficult for future maintainers to follow the callback through the code; with the unwrap it is clear that the author (us) intended for the closure to still be present when we take it from the config.

I see what _you_ mean and I agree with you. My issue comes more from the pattern; that we have to use `Option<Deferred>` to take it, the `Option` reads to me like `Option`al in this case, much like the callback is for `Connection`. 

That's also what got me thinking in terms of a unique pointer instead, if my `Option<Deferred>` is `Some` then I own it and is responsible for dropping it. Just spitballing now but maybe a `type Unique = Option<Deferred>` could work?
Then you could still do: `on_message.take().unwrap()`, only now:

```
pub struct SubscriberConfig {
    ptr: *mut mdb_sys::mdb_subscriber_config_t,
    on_message: Unique,
}
```
I'm not sure, maybe what you did to begin with is the best option.

Reply:
I reverted the change and tried something else. If it looks like an improvement, then maybe I finally understand what you meant this time :grin:

Reply:
I was thinking about the brief moment when it is `None` while `drop` runs, but the type already communicates that technically it is optional, so I like your phrasing better. Fixed.